### DISCUSSION TITLE
What's the rationale for `getQueryData()` not accepting a partial `queryKey`?

### DISCUSSION THREAD

**Discussion by sam-rad** [TanStack/query#6685] [Q&A] [answered]
Looking at the [client api](https://tanstack.com/query/v4/docs/react/reference/QueryClient) if I have some data stored with key `['x', 1]`:

```js
client.getQueryData(["x", 1]);     // this returns data
client.getQueryData(["x"]);        // this returns undefined

client.getQueriesData(["x", 1]);   // this returns data in tuple
client.getQueriesData(["x"]);      // this returns data in tuple
```

Why `getQueriesData` accepts a partial key and properly returns data but `getQueryData` doesn't?

**Comment by vossmalte** [comment, marked as answer]
Imagine: you have data for `['x',1]` and `['x',2]` and `['x']`.
What result expect do you expect from `client.getQueryData(["x"]);`?

`getQueriesData` does accept a partial key as being part of a [filter](https://tanstack.com/query/latest/docs/react/guides/filters).
On the other hand, `getQueryData` looks up a specific cache entry (and if it does not exist, it is `undefined`).

**Reply by sam-rad** [reply]
In case of multiple matches just return the first one? Even if you're against this, it would be very useful to explicitly mention it in the documentation. I'll be happy to do that.

**Reply by vossmalte** [reply]
And which is the "first" one?

**Reply by sam-rad** [reply]
Depends on how the lookup is done. For the plural version you collect a list based on some criteria. For the single version just return after finding the first instance that matched.

**Reply by sam-rad** [reply]
All I'm saying is it's confusing that these two behave differently without any explicit mentions in the docs. Suggesting to fix the confusion there are two options:

 - Make it clear in the documentation
 - Make the api behave the same

**Reply by TkDodo** [reply]
> return the first one

the first one of what, insertion order? It's an implementation detail that we use a `Map` internally. You would also get different result depending on which order the components rendered. Why would this be a good idea, and more importantly - what would it be needed for?

The difference is that `getQueryData` takes a key, and `getQueriesData` takes a `QueryFilter`. Yes, the key can act as a queryFilter, but that's about as far as the similarities go. We've also tried to make sure to reflect this in the naming: Functions that have a plural in their name (`queries` in `getQueriesData`) take a filter, match fuzzily and can return multiple entries. Functions that have singular (`query` in `getQueryData`) take only a key, match exactly and only return one entry or undefined.

The docs are pretty clear on this imo:

https://tanstack.com/query/v5/docs/react/reference/QueryClient#queryclientgetquerydata

> getQueryData is a synchronous function that can be used to get an existing query's cached data. If the query does not exist, undefined will be returned.

vs.

https://tanstack.com/query/v5/docs/react/reference/QueryClient#queryclientgetqueriesdata

> getQueriesData is a synchronous function that can be used to get the cached data of multiple queries. Only queries that match the passed queryFilter will be returned. If there are no matching queries, an empty array will be returned.

**Reply by sam-rad** [reply]
The docs says they both accept `filters`:

![image](https://github.com/TanStack/query/assets/12651119/2a78440e-8fb6-4c40-a4b0-f4cf4e2bdee7)

![image](https://github.com/TanStack/query/assets/12651119/4e432507-78ef-4a3f-bc97-68b0e003e52c)

**Reply by TkDodo** [reply]
not in v5!

**Reply by sam-rad** [reply]
That's what you get when you google the api. So it would be great to clarify that this queryFilter is slightly different from the other queryFilter.
